File: WebClient SDK/src/default/tiles/stageSourceBuildTileExtension.ts
import { sendCompileRequest, CompiledCardTypes } from '../workflow/krProcess/krUIHelper';
import { TileExtension, ITileGlobalExtensionContext, Tile, TileGroups, ITileLocalExtensionContext, TileHotkey } from 'tessa/ui/tiles';
import { CardStoreMode } from 'tessa/cards';

export class StageSourceBuildTileExtension extends TileExtension {

  public initializingGlobal(context: ITileGlobalExtensionContext) {
    const panel = context.workspace.leftPanel;
    const contextSource = panel.contextSource;

    const tile = new Tile({
      name: 'StageSourceBuild',
      caption: '$UI_Tiles_StageSourceBuild',
      icon: 'ta icon-thin-096',
      contextSource,
      command: async () => await sendCompileRequest('.CompileWithValidationResult'),
      group: TileGroups.Cards,
      order: 26,
      evaluating: e =>  {
        const editor = e.currentTile.context.cardEditor;
        e.setIsEnabledWithCollapsing(e.currentTile,
          !!editor
          && !!editor.cardModel
          && CompiledCardTypes.some(x => x === editor.cardModel!.cardType.id
          && editor.cardModel!.card.storeMode === CardStoreMode.Update)
        );
      },
      toolTip: '$UI_Tiles_StageSourceBuild_ToolTip'
    });

    panel.tiles.push(tile);
  }

  public initializingLocal(context: ITileLocalExtensionContext) {
    const leftPanel = context.workspace.leftPanel;
    const tile = leftPanel.tryGetTile('StageSourceBuild');
    if (!tile) {
      return;
    }

    const hotkeyStorage = leftPanel.contextSource.hotkeyStorage;
    hotkeyStorage.addTileHotkey(new TileHotkey(tile, 'Ctrl+Shift+B', 'KeyB', {ctrl: true, shift: true}));
  }

}

File: Source/Extensions/Tessa.Extensions.Default.Client/Tiles/StageSourceBuildTileExtension.cs
using System.Linq;
using System.Threading.Tasks;
using Tessa.Cards;
using Tessa.Extensions.Default.Client.Workflow.KrProcess;
using Tessa.Extensions.Default.Shared.Workflow.KrProcess;
using Tessa.Platform;
using Tessa.UI;
using Tessa.UI.Tiles;
using Tessa.UI.Tiles.Extensions;

namespace Tessa.Extensions.Default.Client.Tiles
{
    public sealed class StageSourceBuildTileExtension: TileExtension
    {
        #region Private Methods

        private static void EnableForSendRequest(object sender, TileEvaluationEventArgs e)
        {
            var model = e.CurrentTile.Context.CardEditor?.CardModel;

            e.SetIsEnabledWithCollapsing(
                e.CurrentTile,
                model != null
                && KrConstants.CompiledCardTypes.Any(p => model.CardType.ID == p)
                && model.Card.StoreMode == CardStoreMode.Update);
        }

        #endregion

        #region base overrides

        public override Task InitializingGlobal(ITileGlobalExtensionContext context)
        {
            var panel = context.Workspace.LeftPanel;
            var tiles = panel.Tiles;

            var tile =
                new Tile(
                    TileNames.StageSourceBuild,
                    "$UI_Tiles_StageSourceBuild",
                    context.Icons.Get("Thin96"),
                    panel,
                    new DelegateCommand(async o => await KrUIHelper.SendCompileRequestAsync(KrConstants.Keys.CompileWithValidationResult)),
                    TileGroups.Cards,
                    order: 26,
                    toolTip: TileHelper.GetToolTip("$UI_Tiles_StageSourceBuild_ToolTip", KrTileKeys.StageSourceBuild),
                    evaluating: EnableForSendRequest);

            tiles.Add(tile);
            return Task.CompletedTask;
        }

        public override Task InitializingLocal(ITileLocalExtensionContext context)
        {
            var tile = context.Workspace.LeftPanel.Tiles.TryGet(TileNames.StageSourceBuild);
            tile?.Context.AddInputBinding(tile, KrTileKeys.StageSourceBuild);
            return Task.CompletedTask;
        }

        #endregion
    }
}